### ISSUE TITLE
"ZeroDivisionError: integer division or modulo by zero" while backpropagating

### ISSUE THREAD

**Issue by LukeBolly** [tensorflow/tensorflow#51653]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below):
tensorflow              2.5.0
tensorflow-estimator    2.5.0
tensorflow-probability  0.12.2

- Python version: 3.8
- CUDA/cuDNN version: 8101
- GPU model and memory: GeForce GTX 1060

**Describe the current behavior**
Trying to train a component of my model throws an error during backprop calculation. The error isn't very clear about what I'm doing wrong. Other parts of the model are training okay, but one section is throwing the following error.

```
    grads = tape.gradient(loss, varibs)
  File "C:\Users\Luke\Anaconda3\envs\ScDreamer\lib\site-packages\tensorflow\python\eager\backprop.py", line 1074, in gradient
    flat_grad = imperative_grad.imperative_grad(
  File "C:\Users\Luke\Anaconda3\envs\ScDreamer\lib\site-packages\tensorflow\python\eager\imperative_grad.py", line 71, in imperative_grad
    return pywrap_tfe.TFE_Py_TapeGradient(
  File "C:\Users\Luke\Anaconda3\envs\ScDreamer\lib\site-packages\tensorflow\python\eager\backprop.py", line 159, in _gradient_function
    return grad_fn(mock_op, *out_grads)
  File "C:\Users\Luke\Anaconda3\envs\ScDreamer\lib\site-packages\tensorflow\python\ops\array_grad.py", line 229, in _ConcatGradV2
    return _ConcatGradHelper(
  File "C:\Users\Luke\Anaconda3\envs\ScDreamer\lib\site-packages\tensorflow\python\ops\array_grad.py", line 119, in _ConcatGradHelper
    concat_dim._numpy().item(0) % input_values[0]._rank())  # pylint: disable=protected-access
ZeroDivisionError: integer division or modulo by zero
```

Investigating the line of code, the issue is the rank of a scalar is being used in the modulo
![image](https://user-images.githubusercontent.com/24449147/130629680-405695f9-642b-4be6-b4c0-acc6e3366267.png)

I'm not sure what I've done wrong that has caused this.

**Describe the expected behavior**
Ideally the error would be caught before getting this deep and suggest a proper fix. At the moment, being told "ZeroDivisionError: integer division or modulo by zero" isn't helping me figure out what is wrong.

**Standalone code to reproduce the issue**
I don't really understand what the issue is here so I'm not sure where to start to try and reproduce this. If I can be pointed in the right direction to reproduce, I'll be happy to write some code.

**Comment by LukeBolly** [comment]
I figured out a workaround, but I'm pretty sure that this is a bug.

The very last step of my loss calculation was concatenating a bunch of scalars together and taking the mean, by changing them from scalars to rank 1 tensors with keep_dims, the error no longer occurs.

some pseudo-code:

```
actor_losses = []
for e in error:
    actor_loss = -tf.reduce_mean(e, axis=0, keepdims=True)        # removing axis & keepdims args will result in scalar output, causing error
    actor_losses.append(actor_loss)
total_loss = tf.reduce_mean(tf.concat(actor_losses, axis=0))
```

**Comment by tilakrayal** [comment]
@LukeBolly ,
 In order to expedite the trouble-shooting process, could you please provide a complete code and the dataset you are using.Also please refer to these links [1](https://github.com/google/emoji-scavenger-hunt/issues/28) and [2](https://stackoverflow.com/questions/44998778/keras-zerodivisionerror-integer-division-or-modulo-by-zero) and let us know if it helped.Thanks!

**Comment by sannamir** [comment]
Hi! I am new at open source contribution. Is this issue still open? I would like to work on it.

**Comment by LukeBolly** [comment]
> Hi! I am new at open source contribution. Is this issue still open? I would like to work on it.

Yep, I probably won't have time to work on this for a while so it's all yours if you like. Let me know if you need any more information to reproduce the issue

**Comment by sannamir** [comment]
@LukeBolly Can you please help me with the review on my PR? It says to add a test for the behavior you're fixing.

**Comment by LukeBolly** [comment]
I think you need to add a bug reproduce into
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/python/kernel_tests
I'm not 100% if bug repros go into kernel_tests, maybe one of the TF team can advise you on where the most appropriate location is

**Comment by LukeBolly** [comment]
What do you have so far? I don't have much time at the moment so I can't write the repro for you, but the gist of the error is that backpropagating through a list of scalars should trigger it.

**Comment by tilakrayal** [comment]
@LukeBolly ,

Please check the above [comment](https://github.com/tensorflow/tensorflow/issues/51653#issuecomment-905196574) and provide the complete code to reproduce the issue.Thanks!

**Comment by sannamir** [comment]
how do I create unit tests?
